### DISCUSSION TITLE
Zod can not have an optional url

### DISCUSSION THREAD

**Discussion by Hasanovic-Kahya** [colinhacks/zod#2801] [Q&A] [answered]
z.string().optional() is not a problem 
but z.string().url().optional() doesnt work
and z.url() doesnt exist 
i can just do it with z.string().regex()

**Comment by hudson-bruno** [comment]
I just tried and it works just fine, what exactly does not work? Just a reminder that if you want to pass null you should use `.nullable()` instead of `.optional()`

**Comment by amit548** [comment]
same problem :(

> z.string().optional() is not a problem but z.string().url().optional() doesnt work and z.url() doesnt exist i can just do it with z.string().regex()

same problem :(

**Comment by colinhacks** [comment, marked as answer]
I'm not sure what is being asked. This is currently working as expected.
```ts
z.string().url().optional().parse(undefined)
// => undefined
```

**Reply by Hasanovic-Kahya** [reply]
thank you

**Comment by JacobWeisenburger** [comment]
@Hasanovic-Kahya or @amit548 
Please send a full reproducible code example.

**Comment by AuthorProxy** [comment]
this is still doesn't work https://codepen.io/AuthorProxy/pen/KKJjxZX

<img width="352" alt="image" src="https://github.com/colinhacks/zod/assets/1407631/ee5cafa1-d89d-4b0e-bded-d50e78873341">
<img width="858" alt="image" src="https://github.com/colinhacks/zod/assets/1407631/c64e3de7-97d8-44b5-9ebb-7eb8997e20f3">

**Reply by AuthorProxy** [reply]
I think it is the same as https://github.com/colinhacks/zod/issues/310

`z.string().url().or(z.literal(''))` - is an answer, but it is awful, plz add internal method for this, don't torture your users, it is the problem every 2nd trying to fix and spent a lot of time. At least add this to docs plz - section form validation.

**Comment by elhasnaouiValinnov** [comment]
Instead of optional what you are probably after is a union with the empty string?
 z.union([z.literal(""), z.string().trim().url()])
